File: solution/473/solution.py
class Solution(object):
    def makesquare(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        if len(nums) < 4:return False
        nums.sort(reverse=True)
        num_s = sum(nums)
        if num_s % 4 != 0:return False
        target = num_s / 4
        if nums[0] < target and nums[0] + nums[-1] > target:return False
        backet = [0] * 4
        return self.generate(0, backet, target, nums)

    def generate(self, i, backet, target,nums):
        if i >= len(nums):
            return backet[0] == target and backet[1] == target and backet[2] == target and backet[3] == target
        for x in xrange(len(backet)):
            backet[x] += nums[i]
            if backet[x] <= target and self.generate(i+1, backet, target, nums):
                return True
            backet[x] -= nums[i]
        return False

s = Solution()
print s.makesquare([1,1,2,2,2])

File: solution/473/solution.cpp
class Solution {
public:
    bool makesquare(vector<int>& nums) {
        if(nums.size()<4)return false;
        sort(nums.rbegin(), nums.rend());
        int sum = 0;
        for(int i=0;i<nums.size();i++)sum+=nums[i];
        if(sum%4!=0)return false;
        int target = sum/4;
        int items[4] = {0};
        return generate(0, items, target, nums);
    }
    bool generate(int n, int items[], int target, vector<int>& nums){
        if(n>=nums.size())return items[0] == target&&items[1] == target&&items[2] == target&&items[3] == target;
        for(int i=0;i<4;i++){
            items[i] += nums[n];
            if(items[i]<=target){
                if(generate(n+1,items,target,nums))return true;
            }
            items -= nums[n];
        }
        return false;
    }
};